pkg/util/webhooks/validating-webhooks/validating-webhook.go

```diff
@@ -93,3 +110,23 @@ func Serve(resp http.ResponseWriter, req *http.Request, admitter admitter) {
 		return
 	}
 }
+
+func getContextFromRequest(req *http.Request) (context.Context, context.CancelFunc, error) {
+	const defaultTimeout = 10 * time.Second
+	timeout := defaultTimeout
+
+	if timeoutStr := req.URL.Query().Get("timeout"); timeoutStr != "" {
+		parsedTimeout, err := time.ParseDuration(timeoutStr)
+		if err != nil {
+			errTxt := fmt.Sprintf("failed to parse timeout duration of %q", timeoutStr)
+			log.Log.Reason(err).Error(errTxt)
+			return nil, nil, fmt.Errorf(errTxt)
```

Comment:
What is the reason to fail in this case?
Can't we silently fallback to the default?

Reply:
no - this is a wrong user input. we must not handle these kind of mistakes

Reply:
> once you expose an API - you can't guess who'll use it

I don't follow, could you please elaborate?

The way I see it, `virt-api` is a dedicated server that has several functionalities.
One of its functionalities is serving webhooks - which is done using a [dedicated API](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#webhook-request-and-response).

`virt-api` resides in a dedicated namespace (that should be unreachable to regular users) and is being exposed by a dedicated [service](https://github.com/kubevirt/kubevirt/blob/3dbf033c27af8d39749e8360dba81d79478aa89e/pkg/virt-operator/resource/generate/components/deployments.go#L89) that AFAIU is only meant to serve `kube-api`.

At the moment, I cannot think of another legitimate component / user trying to query virt-api's webhook 
endpoints other than `kube-api`.

A hard requirement for expecting the `timeout` query string basically changes our API with `kube-api` and will cause `virt-api` to break in case it will be renamed or dropped in the future (although unlikely).

Not having a hard requirement for the `timeout` query string has several advantages:
1. Less code paths.
2. Less error handling.
3. The cancel function will always be called unconditionally - making the code more predicable and requires less mental energy to follow.
4. There is no need for a dedicated unit tests for the context logic.

WDYT?

Reply:
This is not exactly the case. Bottom line, virt-api is just a HTTP server. Even if you wrap it with clients, it is still a HTTP endpoint that can get request from everywhere. You can't really control it.

Reply:
What's wrong with error handling? and unit tests?

Reply:
About the defer, I just forgot to simplfy it - pushing a fix.

Reply:
On both alternative there are three layers of defense:
1. The context from the http request.
2. The value from timeout.
3. Default timeline value.

So the argument is between two similar level of protection, but different levels of complexity.

I would also like to raise the "Robustness principle" [1] as an argument:
"Be conservative in what you do, be liberal in what you accept from others".

[1] https://en.wikipedia.org/wiki/Robustness_principle

Reply:
This is not related to timeout but for bad request. If the request is wrong, we must not handle it

Reply:
> This is not related to timeout but for bad request. If the request is wrong, we must not handle it

Fair enough.